Run automated test of Sage Install Script

jazzsequence opened the pull request:
This PR creates a new GH action workflow to test the Sage Theme install script (in `./private/scripts/helpers.sh`) on MacOS and Linux environments (`ubuntu-latest` and `macos-latest`). (Windows tests were attempted but ultimately abandoned due to issues in Terminus and a [lack of support for Windows Terminus installs](https://github.com/pantheon-systems/terminus#ubuntu--winwslubuntu).)

The matrix of tests has been configured to use PHP 8.x versions on each of the OSes it's being tested on (Roots Bedrock does not support < PHP 8).

In addition, changes to the script has been made so it can be run without user input via environment variables. These environment variables can be used to define any of the user input fields as well as whether the script is being run by a CI.

Future improvements could include Windows support (pending official support in the Terminus project), ~~an environment variable for PHP version (so each PHP version in the matrix is actually updated so the created multidev is using that PHP version)~~ (this has been rolled in) ~~and documentation for using the environment variables~~ (documentation has been added), and testing on WordPress multisite.

closes #112

jazzsequence committed:
adds a sage-test workflow

jazzsequence committed:
temporarily allow the test to run on any branch so we can see if it works

jazzsequence committed:
add to on push to test the test

jazzsequence committed:
pull terminus.phar from latest

jazzsequence committed:
authenticate with terminus

jazzsequence committed:
this is a github secret, not an environment variable

jazzsequence committed:
create a multidev for the build in a separate step
need to add a no-input parameter for the install script

jazzsequence committed:
add a step to generate a name for our multidev
based on the os, pr, and php version

jazzsequence committed:
use github.event.number

jazzsequence committed:
run a cleanup step that deletes the multidev

jazzsequence committed:
add an echo so we can save the output of the pr number

jazzsequence committed:
get the pr number a different way?

jazzsequence committed:
remove core const if already declared

jazzsequence committed:
echo the PR number?

jazzsequence committed:
echo all the things

jazzsequence committed:
use gh cli?

jazzsequence committed:
get the pr number the right way

jazzsequence committed:
explicitly set the shell to bash everywhere

jazzsequence committed:
set the default shell to bash and remove the definitions

jazzsequence committed:
if we're in a windows environment, how we install terminus differs
add the current directory to the PATH variable so we can run terminus from where we downloaded it

jazzsequence committed:
Sage (and Bedrock) doesn't actually work on PHP 7.4

jazzsequence committed:
remove windows tests
terminus isn't technically supported on windows so this is kind of moot

jazzsequence committed:
install terminus from homebrew if macos

jazzsequence committed:
add a test that terminus works

jazzsequence committed:
add latest version back

jazzsequence committed:
shorten the test name

jazzsequence committed:
support values to be globally set

jazzsequence committed:
globally set values
we might not actually need this if we're in the site directory...

jazzsequence committed:
clone down the site
can't believe I went this long without realizing we're going to need to run this in an actual site directory

jazzsequence committed:
sync the latest changes to the test multidev

jazzsequence committed:
wrap sitename in quotes

jazzsequence committed:
install ssh key so we can clone

jazzsequence committed:
clone site before multidev creation
if it's going to fail, let's fail before we wait for the multidev to be created

jazzsequence committed:
deal with pantheon ssh host keys

jazzsequence committed:
remove the ssh-keyscan step

jazzsequence committed:
need to set a git config before we can git things

jazzsequence committed:
make sure we're in the right directory

jazzsequence committed:
add a check for CI
so we don't loop through the script

jazzsequence committed:
export our globals
so they are read by the script

jazzsequence committed:
don't prompt for confirmation if ci

jazzsequence committed:
wait longer for CI since there are concurrent builds

jazzsequence committed:
move these to be global variables and add a siteenv variable

jazzsequence committed:
use siteenv anywhere we'd manually set .dev

jazzsequence committed:
define the SITEENV to the multidev name

jazzsequence committed:
lower the waittime to 300
i think 600 might be too much

jazzsequence committed:
wait for long running scripts but wait shorter
but still vary it based on whether we're on ci or local

jazzsequence committed:
add colors

jazzsequence committed:
output that we're checking for jq

jazzsequence committed:
allow sed handing between linux and macos and bail if it fails

jazzsequence committed:
fix sed condition for shellcheck

jazzsequence committed:
print current working directory

jazzsequence committed:
sed conditional needs to go inside the ostype condition

jazzsequence committed:
add closing fi

jazzsequence committed:
add a check for the sage theme directory

jazzsequence committed:
move the git pull to after we push the post install cmd
i think this isn't actually getting pushed

jazzsequence committed:
remove random + that must've been a typo

jazzsequence committed:
sitename.siteenv

jazzsequence committed:
halve the waittime yet again

jazzsequence committed:
push to the right branch

jazzsequence committed:
only try opening the site if we're not in ci

jazzsequence committed:
wrap $branch in quotes

jazzsequence committed:
remove composer install step
we do this again later inside the multidev that we clone down, so there's no reason to do this before we create the multidev

jazzsequence committed:
break terminus stuff up into multiple steps

jazzsequence committed:
delete sage theme if it exists

jazzsequence committed:
composer install after copying updates from source

jazzsequence committed:
git diff before create multidev

jazzsequence committed:
git status instead of git diff
bail if dirty

jazzsequence committed:
temporarily gitignore sage test to create multidev

jazzsequence committed:
delete all tags from the CI site

jazzsequence committed:
delete the build artifact tag

jazzsequence committed:
remove the gitignore steps
we don't need them
or we shouldn't, anyway

jazzsequence committed:
fetch the tags before deleting the multidev artifact

jazzsequence committed:
remove extra git pushes
so we have less sync code failures

jazzsequence committed:
underscore in the build artifact tag name

jazzsequence committed:
the other underscore, too

jazzsequence committed:
allow other versions of PHP

jazzsequence committed:
change output from listing to checking
we don't output the themes dir, so we shouldn't sey we're listing it

jazzsequence committed:
add 8.3

jazzsequence committed:
allow the build artifact tag deletion to fail
if it doesn't exist, that's okay

jazzsequence committed:
actually bump the PHP version to the version declared

jazzsequence committed:
add a note when checking for unfinished/long-running workflows

jazzsequence committed:
allow both tag deletes to fail

jazzsequence committed:
remove pwd

jazzsequence committed:
set phpversion to 8.0 if 8 is passed

jazzsequence committed:
do a git status to see the change

jazzsequence committed:
bump version of ssh-agent

jazzsequence committed:
set output will be deprecated so use env variables instead
https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

jazzsequence committed:
get and set in the same step

jazzsequence committed:
remove the run of this workflow on every push
just need it on merges and when we make changes to the scripts or this workflow

jazzsequence committed:
run the workflow whenever updates are pushed the test itself

jazzsequence committed:
add documentation for running the script headless

jazzsequence committed:
update path to script

jazzsequence committed:
update docs about CI env variable

jazzsequence committed:
remove extra line

jazzsequence committed:
update changelog

jazzsequence committed:
remove preceding slash for path to scripts

pwtyler commented on .github/workflows/sage-test.yml:
```diff
@@ -0,0 +1,149 @@
+      - name: Get and Set PR number
+        id: pr
+        run: |
+          pr_num=$(gh pr view --json number -q .number || echo "")
+          echo "pr_num=$pr_num" >> $GITHUB_ENV
```
nit: env variables should be all caps

jazzsequence replied:
yeah, but if I do that, it wouldn't match with the other variables on line 55-57 (https://github.com/pantheon-systems/wordpress-composer-managed/pull/113/files#diff-2faa71adfe1b6b12a9b18a4d722c7463005f533d54912c2154589a1d6b91b3bcR55-R57) or I would need to do `pr_num=$(echo $PR_NUM)`

jazzsequence committed:
fix spacing

pwtyler commented on .github/workflows/sage-test.yml:
```diff
@@ -0,0 +1,149 @@
+            brew install pantheon-systems/external/terminus
+          else
+            echo "💻 Installing Terminus from phar on Linux..."
+            curl -L "https://github.com/pantheon-systems/terminus/releases/download/$latest_version/terminus.phar" -o terminus
```
github should have a latest release endpoint so you needn't look it up.

Optional.

pwtyler commented on .github/workflows/sage-test.yml:
```diff
@@ -0,0 +1,149 @@
+          export SITEENV=$multidev_name
+          export PHPVERSION=${{ matrix.php-version }}
+          cd ~/pantheon-local-copies/$SITENAME
+          composer install-sage || exit_code=$?
+          if [[ -n "${exit_code}" ]]; then
+            echo "❌ Sage Install Script Failed with exit code ${exit_code}"
+            exit $exit_code
+          else
+            echo "✅ Sage Install Script passed!"
+            exit 0;
+          fi
```
why check the exit code? Won't CI bail if this exits non-zero?
```suggestion
          composer install-sage
          echo "✅ Sage Install Script passed!"
```

pwtyler commented on .github/workflows/sage-test.yml:
```diff
@@ -0,0 +1,149 @@
+            echo "✅ Sage Install Script passed!"
+            exit 0;
+          fi
+      - name: Delete multidev
+        if: always()
+        run: terminus multidev:delete --delete-branch --yes wpcm-sage-install-tests.$multidev_name
```
Is multidev env not useful to review in case of failed tests? Seems should should only run on success, and otherwise have validation _in front_ of the CI that ensures there are enough environments.

jazzsequence replied:
I don't know that there's anything the multidev would tell you (that you couldn't get from workflow logs) that you wouldn't see in the output of the automated run. What's more, because the multidev environment is named based on the PR #, it will always create a multidev named `${os}-${php_version}-${pr_num}` and then we'd need to do all sorts of cleanup on the environment for subsequent runs on the same PR.

pwtyler commented on private/scripts/helpers.sh:
```diff
@@ -54,6 +68,10 @@ function get_info() {
 
   # Unset the variables if we're doing this a second time.
   if [ "$is_restarted" -eq 1 ]; then
+    if [ "$is_ci" -eq 1 ]; then
```
`-eq` might be unhappy if compared to `""`? Default to 0 up on line 23, or test with -n/-z first or just use [[ == ]] instead.

pwtyler commented on private/scripts/helpers.sh:
```diff
@@ -138,15 +156,16 @@ function get_info() {
+    fi
   fi
 
   if [ -z "$sitename" ] || [ -z "$sagename" ] || [ -z "$sftpuser" ] || [ -z "$sftphost" ]; then
```
I can't comment on the line below (172) but `"echo "${red}Missing information!${normal} Make sure you've everything for all the prompts.""` needs a typo fix.

pwtyler commented on private/scripts/helpers.sh:
```diff
@@ -230,13 +264,14 @@ function update_php() {
     else
       # Add full PHP version declaration to pantheon.yml.
-      echo "php_version: 8.0" >> pantheon.yml
+      echo "php_version: ${phpversion}" >> pantheon.yml
```
```suggestion
      echo "" >> pantheon.yml
      echo "php_version: ${phpversion}" >> pantheon.yml
```
learned this the hard way when there's no newline at the end of the file...

pwtyler commented on private/scripts/helpers.sh:
```diff
@@ -230,13 +264,14 @@ function update_php() {
     fi
-    git commit -am "[Sage Install] Update PHP version to 8.0"
-    git push origin master
+    git status
```
```suggestion
```
still useful?

jazzsequence replied:
No because what I actually wanted was to see what version actually got added to the `pantheon.yml`. What I actually want is a `git diff`.

pwtyler commented on private/scripts/helpers.sh:
```diff
@@ -359,47 +391,78 @@ function update_composer() {
   # Wait for the build to finish.
   echo "${yellow}Waiting for the deploy to finish.${normal}"
-  if ! terminus workflow:wait --max=90 "$sitename".dev; then
+  if [ "$is_ci" -eq 1 ]; then
```
same as above.

pwtyler commented on private/scripts/helpers.sh:
```diff
@@ -359,47 +391,78 @@ function update_composer() {
 
 # Finish up the Sage install process.
 function clean_up() {
+  # List the app/ehemes directory.
```
```suggestion
  # List the app/themes directory.
```

pwtyler commented on private/scripts/helpers.sh:
```diff
@@ -408,28 +471,32 @@ function clean_up() {
-  open https://dev-"$sitename".pantheonsite.io
+  if [ "$is_ci" -ne 1 ]; then
+    # Open the site. This should generate requisite files on page load.
+    echo "${yellow}Opening the dev-${sitename}.pantheonsite.io to generate requisite files.${normal}"
+    open https://dev-"$sitename".pantheonsite.io
```
should this be `$siteenv` not dev?

jazzsequence committed:
use latest terminus rather than figuring out a version

jazzsequence committed:
Don't worry about checking the exit code

Co-authored-by: Phil Tyler <philip@tylerdigital.com>

jazzsequence committed:
add a newline before php_version

Co-authored-by: Phil Tyler <philip@tylerdigital.com>

jazzsequence committed:
default to 0

jazzsequence committed:
Merge branch '112-try-testing-sage-install-script' of github.com:pantheon-systems/wordpress-composer-managed into 112-try-testing-sage-install-script

jazzsequence committed:
fix missing info prompt

jazzsequence committed:
git diff, not status

jazzsequence committed:
fix typo

Co-authored-by: Phil Tyler <philip@tylerdigital.com>

jazzsequence committed:
use $siteenv

jazzsequence committed:
use env varibales

jazzsequence committed:
move the path filter for the scripts folder to push
instead of PR, which doesn't seem to be triggering

jazzsequence committed:
wrap multidev_name in quotes

jazzsequence committed:
export SITEENV inside the script
$multidev_name is not carrying over

jazzsequence commented on .github/workflows/sage-test.yml:
```diff
@@ -0,0 +1,149 @@
+          export SITENAME=wpcm-sage-install-tests
+          export CI=1
+          export SITEENV=$multidev_name
+          export PHPVERSION=${{ matrix.php-version }}
```
This works for everything except for `SITEENV` because the `GITHUB_ENV` variable `$multidev_name` doesn't get carried over into environment variables...maybe there's some ${{ github.env }} thing that could be used...

jazzsequence committed:
use multidev_name from env

jazzsequence committed:
all caps env variables

jazzsequence commented on .github/workflows/sage-test.yml:
```diff
@@ -0,0 +1,149 @@
+        id: pr
+        run: |
+          pr_num=$(gh pr view --json number -q .number || echo "")
+          echo "pr_num=$pr_num" >> $GITHUB_ENV
```
leaving because pulling the PR number directly from gh is simpler parsing out of GITHUB_REF

jazzsequence committed:
don't switch to sftp mode
since we're not opening the window, we don't need to switch to sftp and then switch back

jazzsequence committed:
actually we can still commit leftover files
just don't open the browser
